metric: add cidr routing to backup and external storage

andrewbaptist opened the pull request:
Add CIDR based lookup cache and modify the sql metrics for bytes in and out to become agg metrics rather than regular counter metrics. 

Epic: none

Release note: 
Modifies metrics `sql.bytesin` and `sql.bytesout` to be agg metrics if the child metrics are enabled. 
Adds two new configuration parameters: `metrics.cidr_mapping_url` and `metrics.cidr_mapping_transform`

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126891)
<!-- Reviewable:end -->

Commit:
util: add utility to process CIDR records

This commit adds a new utility which can store and efficiently process a
large number of CIDR records by mapping them to a unique name.

Epic: CRDB-41142

Release note: None

Commit:
sql: restructure sql byte metrics to be add metrics

This commit changes the SQL byte metrics to be broken down by cidr block
of the source.

Epic: CRDB-41142

Release note (ops change): Modifies metrics sql.bytesin and sql.bytesout
to be agg metrics if child metrics are enabled.

Commit:
sql: propagate the cidr.Lookup into ExecutorConfig

This commit constructs the cidr.Lookup into the sql server
ExecutorConfig and evalContext. Additionally this enables the cidr
mapping and adds a new configuration parameter for it.

Epic: CRDB-41142

Release note (ops change): Adds a new configuration parameters
server.cidr_mapping_url which maps IPv4 CIDR blocks to arbitrary tag
names.

adwittumuluri commented on pkg/util/cidr/cidr.go:
```diff
@@ -0,0 +1,297 @@
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+)
+
+var cidrMappingUrl = settings.RegisterStringSetting(
+	settings.ApplicationLevel,
+	"metrics.cidr_mapping_url",
+	"url of a JSON file containing a list of CIDR blocks (file:// or http://)",
+	envutil.EnvOrDefaultString("COCKROACH_CIDR_MAPPING", ""),
+	settings.WithPublic,
+)
```
Will it be possible to declare multiple `cidr_mapping_url`s?

* If we go the route of pointing CRDB directly to each cloud's IP range list, then we'll have three different URLs to declare. Here, if we _don't_ support multiple files, then CC will never need to use `cidr_mapping_transform`, and would go with the local file route.
* If we go the route of pointing CRDB to a local file, it may happen that that file is too large to (easily) propagate it to the nodes file system. Granted, we can probably work around the latter restriction ([thread](https://cockroachlabs.slack.com/archives/C07597E1JGZ/p1720638878877559)), but wanted to check

andrewbaptist replied:
It would be possible to have multiple URLs, but I wasn't sure if this is the right approach. I also agree we probably won't end up using the transform, and it would be easy to remove it. It made some of the testing easier, but other than that, if we go the file route it doesn't make much difference. 

There are various ways we can do this in the end, initially I was trying to add enough flexibility to allow you to try out some different ways and we can clean it up before we merge the PR.

In terms of multiple files, that would also be possible. An alternative is if the pointed to file is a directory (or regex), the code could scan the directory (or regex) and load all files that match. 

All of these options are pretty easy and self-contained in that one commit, so I'm not too concerned about changing it.

adwittumuluri replied:
I'm in favor of pointing to a single file in GCS. It seems less complicated than multiple files, compression, etc. I'll work towards that and see how things shake out

adwittumuluri commented on pkg/sql/pgwire/server.go:
```diff
@@ -854,6 +884,21 @@ func (s *Server) ServeConn(
 }
 
+func (s *Server) lookup(ctx context.Context, ip net.IP) destinationMetrics {
+	destination := s.execCfg.IpLookup(ip)
```
I think we discussed following the linux IP tables pattern of applying the most specific cidr-tag rule. I'd like to add a `0.0.0.0/32,public` mapping to catch all non-local, non-cloud addresses. Does this implementation support that?

andrewbaptist replied:
Yes - if we add that rule at the end, it will automatically match as you describe.

Commit:
backupccl: break read and write metrics by bucket

Previously the read and write metrics were aggregate across all
containers or buckets. In some situations it is useful to break the
statistics out to detemine how much data is going to each. This PR
allows breaking the data out for the three main clouds.

Epic: CRDB-41142

Release note: None

Commit:
changefeed: add metrics broken down by destination

This commit changes some of the existing changefeed metrics to be broken
down by CIDR routing block in addition to by job_id.

Epic: CRDB-41142

Release note: None

andrewbaptist commented on pkg/util/cidr/cidr.go:
```diff
@@ -0,0 +1,305 @@
+}
+
+// update every minute if the contents change
+const refreshInterval = 10 * time.Second
```
Change back to 1 minute.

andrewbaptist replied:
Added a config setting, and by default disabled the refresh loop.

andrewbaptist commented on pkg/util/cidr/cidr.go:
```diff
@@ -0,0 +1,305 @@
+		c.refresh(ctx)
+	})
+
+	if err := stopper.RunAsyncTask(ctx, "cidr-refresh", func(ctx context.Context) {
```
Disable auto refresh.

andrewbaptist commented on pkg/cloud/metrics.go:
```diff
@@ -128,19 +144,22 @@ func MakeMetrics() metric.Struct {
 		CreatedReaders: metric.NewCounter(cloudReaders),
 		OpenReaders:    metric.NewGauge(cloudOpenReaders),
-		ReadBytes:      metric.NewCounter(cloudReadBytes),
+		ReadBytes:      aggmetric.NewCounter(cloudReadBytes, "cloud", "container", "cidr"),
```
Rename cidr to label

andrewbaptist commented on pkg/cloud/cloudpb/external_storage.go:
```diff
@@ -64,6 +66,29 @@ func (m *ExternalStorage) AccessIsWithExplicitAuth() bool {
+	case ExternalStorageProvider_azure:
+		return MetricKey{"azure", m.AzureConfig.Container, cidr}
+	}
+	return MetricKey{}
```
Fix to pass 3 values.

Commit:
sql: pass an Inc function to writeBuffer

Previously the writeBuffer required a specific Metric implementation as
part of its paramater. This made it more complicated to change the
metric type that was passed in.

Epic: none

Release note: None
